File: solutions/160.cs
namespace Problem160
{
	public class ListNode
	{
		public int val;
		public ListNode next;
		public ListNode(int x) { val = x; }
	}
	public class Solution
	{
		public ListNode GetIntersectionNode(ListNode headA, ListNode headB)
		{
			// 一定不存在
			if (headA == null || headB == null)
				return null;

			HashSet<ListNode> set = new();
			ListNode currA = headA;

			while (currA != null)
			{
				set.Add(currA);
				currA = currA.next;
			}

			ListNode currB = headB;
			while (currB != null)
			{
				if (set.Contains(currB))
					return currB;
				currB = currB.next;
			}

			return null;
		}
	}
}

File: solutions_cpp/160.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#include <cstddef>
struct ListNode
{
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};
class Solution
{
  public:
    int _getLen(ListNode* head)
    {
        int len = 0;

        while (head != nullptr)
        {
            ++len;
            head = head->next;
        }

        return len;
    }
    ListNode* getIntersectionNode(ListNode* headA, ListNode* headB)
    {
        int lenA = _getLen(headA);
        int lenB = _getLen(headB);
        int deltaLen = 0;

        ListNode* longerHead = nullptr;
        ListNode* shorterHead = nullptr;

        if (lenA >= lenB)
        {
            longerHead = headA;
            shorterHead = headB;
            deltaLen = lenA - lenB;
        }
        else
        {
            longerHead = headB;
            shorterHead = headA;
            deltaLen = lenB - lenA;
        }

        while (deltaLen > 0 && longerHead != nullptr)
        {
            longerHead = longerHead->next;
            --deltaLen;
        }

        while (longerHead != nullptr && shorterHead != nullptr)
        {
            if (longerHead == shorterHead)
            {
                return longerHead;
            }

            longerHead = longerHead->next;
            shorterHead = shorterHead->next;
        }

        return nullptr;
    }
};